### PR TITLE
Directly rewrite filters on RHS join columns into LHS equivalents

### PR DESCRIPTION
**Pull request by jon-wei**
This PR allows `JoinFilterAnalyzer` to directly rewrite filters on RHS join columns into filters on the equivalent LHS join columns, allowing these filters to be pushed down to base tables.

To supports this two new methods are added to `Filter`:
- `supportsRequiredColumnRewrite`
- `rewriteRequiredColumns`

These methods are used to pass a map of column name rewrites to a Filter, which the Filter will use to return a copy of itself that operates on different columns.

Current restrictions on such rewrites:
- Only filters that reference a single column currently support such rewrites
- Expression, Javascript, AND/OR, and Spatial filters are not currently supported
  - For expression and javascript rewrites, we would need a good way to replace the identifiers for the new filter
 
This PR has:
- [x] been self-reviewed.
- [ ] added documentation for new or modified features or behaviors.
- [x] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/licenses.yaml)
- [x] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [x] added unit tests or modified existing tests to cover new code paths.
- [ ] added integration tests.
- [ ] been tested in a test Druid cluster.

### CODE REVIEW

File: processing/src/main/java/org/apache/druid/segment/join/filter/JoinFilterAnalyzer.java [hunk 1]
```diff
@@ -265,6 +286,10 @@ public static JoinFilterPreAnalysis computeJoinFilterPreAnalysis(
                   (rhsCol) -> Optional.of(new ArrayList<>())
               );
           perColumnCorrelations.get().add(correlationForPrefix.getValue());
```
Comment (suneet-s) [inline processing/src/main/java/org/apache/druid/segment/join/filter/JoinFilterAnalyzer.java:288]: get() without isPresent() check

File: processing/src/main/java/org/apache/druid/segment/join/filter/JoinFilterAnalyzer.java [hunk 2]
```diff
@@ -265,6 +286,10 @@ public static JoinFilterPreAnalysis computeJoinFilterPreAnalysis(
...
+          if (rhsRewriteCandidate.isJoinColumn()) {
+            // we don't need to determine correlated values if the filter is on the join column
```
Comment (suneet-s) [inline processing/src/main/java/org/apache/druid/segment/join/filter/JoinFilterAnalyzer.java:290]: Shouldn't we need to do this?

What if the filter could not be re-written directly (eg. an expression filter, or there were multiple columns in a filter)

File: processing/src/main/java/org/apache/druid/query/filter/ExpressionDimFilter.java [hunk 3]
```diff
@@ -44,6 +45,9 @@
   @Nullable
   private final FilterTuning filterTuning;
 
+  @JsonIgnore
+  private final ExprMacroTable macroTable;
```
[the remark this answers is not in the document]
Comment (jon-wei) [reply processing/src/main/java/org/apache/druid/query/filter/ExpressionDimFilter.java:49]: The ExprMacroTable is not really a property of the filter itself, it's an injected reference to a global scoped macro definition table.

It's similarly excluded in aggregator spec equals (e.g. `DoubleSumAggregatorFactory`)

File: processing/src/main/java/org/apache/druid/query/filter/LikeDimFilter.java [hunk 4]
```diff
@@ -363,5 +363,26 @@ public SuffixMatch getSuffixMatch()
     {
       return suffixMatch;
     }
+
```
[the remark this answers is not in the document]
Comment (jon-wei) [reply processing/src/main/java/org/apache/druid/query/filter/LikeDimFilter.java:366]: Like filters can, the new interface is on `Filter`, not `DimFilter`

File: processing/src/main/java/org/apache/druid/query/filter/LikeDimFilter.java [hunk 5]
```diff
@@ -363,5 +363,26 @@ public SuffixMatch getSuffixMatch()
...
+      return getSuffixMatch() == that.getSuffixMatch() &&
+             Objects.equals(getPrefix(), that.getPrefix()) &&
+             Objects.equals(pattern.toString(), that.pattern.toString());
+    }
+
+    @Override
+    public int hashCode()
+    {
+      return Objects.hash(getSuffixMatch(), getPrefix(), pattern.toString());
+    }
```
[the remark this answers is not in the document]
Comment (jon-wei) [reply processing/src/main/java/org/apache/druid/query/filter/LikeDimFilter.java:386]: Hm, it doesn't look like we have EqualsVerifierTest for any of the `DimFilter` implementations; since this patch is only adding equals implementations for `Filter` classes, I think that should be done in a separate PR

File: processing/src/main/java/org/apache/druid/query/filter/ExpressionDimFilter.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (jon-wei) [reply processing/src/main/java/org/apache/druid/query/filter/ExpressionDimFilter.java]: Actually this change is unnecessary (I made it while trying out ExpressionFilter rewrites which i later decided wouldn't be supported initially), I reverted this area

File: processing/src/main/java/org/apache/druid/segment/join/filter/JoinFilterAnalyzer.java (hunk 1 above)
Comment (jon-wei) [reply processing/src/main/java/org/apache/druid/segment/join/filter/JoinFilterAnalyzer.java:290]: I added an assert for `perColumnCorrelations.isPresent()` here, it should never be empty in that block (for it to have been empty, `if (correlationsForPrefix.isPresent()) {` must have been false, and this block ensures it's not empty in the `computeIfAbsent` call)

I think this area of the code is hard to follow right now, it would be an area to be refactored.

File: processing/src/main/java/org/apache/druid/segment/join/filter/JoinFilterAnalyzer.java (hunk 2 above)
Comment (jon-wei) [reply processing/src/main/java/org/apache/druid/segment/join/filter/JoinFilterAnalyzer.java:292]: Hm, I think that `isJoinColumn ` property is more correctly called `isDirectRewrite`, I renamed that.

It's not necessary since if the direct rewrite couldn't happen, then it should fall through to the blocks in `computeJoinFilterPreAnalysis` and  `analyzeJoinFilterClause` which handle indirect rewrites, and it would be handled in that way.

There was a bug that would cause this to not fall through because of missing `supportsRequiredColumnRewrite` checks, I've fixed that and added a filter to `JoinFilterAnalyzerTest.test_filterPushDown_factToRegionToCountryLeftFilterOnRHSJoinConditionColumnsHelper` which overrides `supportsRequiredColumnRewrite` on a SelectorFilter to demonstrate this.

**Comment by jon-wei** [comment]
> I think it's a little hard to follow what's going on in JoinFilterAnalyzer, the class is now up to 1000+ lines, could you re-structure it into smaller classes that are linked to each other to make it a little easier to understand the dependencies / flow in the class.

I definitely agree that's a good idea, I'll do that in a follow-on PR though, since I find that including such refactors in a PR together with functional changes makes it hard to identify the functional changes.

File: processing/src/main/java/org/apache/druid/query/filter/Filter.java [hunk 6]
```diff
@@ -162,4 +163,29 @@ default boolean canVectorizeMatcher()
    * can be expected to have a bitmap index retrievable via {@link BitmapIndexSelector#getBitmapIndex(String)}
    */
   Set<String> getRequiredColumns();
+
+  /**
+   * Returns true is this filter is able to return a copy of this filter that is identical to this filter except that it
+   * operates on different columns, based on a renaming map.
+   */
+  default boolean supportsRequiredColumnRewrite()
```
[the remark this answers is not in the document]
Comment (jihoonson) [reply processing/src/main/java/org/apache/druid/query/filter/Filter.java:171]: Would you elaborate more details of the test you think? I'm not sure how it could help. Since the default returns false, the new filter overrode this method to return true if it had to be. The author should be able to think what it means when it is overridden. If you mean new filters need a design review, that should be done during the design phase rather than testing phase.

File: processing/src/main/java/org/apache/druid/query/filter/Filter.java [hunk 7]
```diff
@@ -162,4 +163,29 @@ default boolean canVectorizeMatcher()
...
+   *
+   * For example, if I have a filter (A = hello), and I have a renaming map (A -> B),
+   * this should return the filter (B = hello)
+   *
+   * @param columnRewrites Column rewrite map
+   * @return Copy of this filter that operates on new columns based on the rewrite map
+   */
+  default Filter rewriteRequiredColumns(Map<String, String> columnRewrites)
+  {
+    throw new UnsupportedOperationException("Required column rewrite is not supported by this filter.");
```
[the remark this answers is not in the document]
Comment (jihoonson) [reply processing/src/main/java/org/apache/druid/query/filter/Filter.java:189]: Good question. A potential alternative could be having just this interface and its default implementation returns `this`. I haven't thought enough to say what is better yet.

File: processing/src/main/java/org/apache/druid/query/filter/Filter.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (suneet-s) [reply processing/src/main/java/org/apache/druid/query/filter/Filter.java:171]: I think that can be difficult for the author and the reviewer. There are many functions in an interface, and most editors will pull in methods that don't have a definition unless you explicitly tell them to. As a reviewer, nothing is warning me that the implementation/ design did not consider implementing this function.

I have a WIP change that does something similar to validate that the equals and hashCode implementations are not using the implementation provided by `Object` I suspect we'd want something very similar, except with the ability to provide exceptions. https://github.com/suneet-s/druid/blob/b8bc17551361b20458b073c62b5fe8f5d6a85183/processing/src/test/java/org/apache/druid/query/filter/FilterTest.java#L33

This way, the dev has to add the exception as part of the PR for the tests to pass, and the reviewer can see that an exception was added and ask themselves if that exception makes sense.

File: processing/src/main/java/org/apache/druid/segment/join/filter/JoinFilterAnalyzer.java [hunk 8]
```diff
@@ -387,87 +430,178 @@ private static JoinFilterAnalysis analyzeJoinFilterClause(
...
+              pushDownVirtualColumn.getOutputName()
+          ));
+          newFilters.add(rewrittenFilter);
+        }
+      }
+    }
+
+    if (newFilters.isEmpty()) {
+      return JoinFilterAnalysis.createNoPushdownFilterAnalysis(filterClause);
+    }
```
Comment (jihoonson) [inline processing/src/main/java/org/apache/druid/segment/join/filter/JoinFilterAnalyzer.java:521]: Hmm, I'm not sure exactly what is wrong, but it seems there is something missing. When I modified the [join condition in this test](https://github.com/apache/druid/blob/master/processing/src/test/java/org/apache/druid/segment/join/JoinFilterAnalyzerTest.java#L1832) to `"\"%sregionIsoCode\" == regionIsoCode && \"%scountryIsoCode\" == regionIsoCode && \"%sregionName\" == user"`, this block did strange rewrites such as `r1.regionIsoCode = AAAA` -> `user = AAAA` or `r1.regionName = Fourems Province` -> `regionIsoCode = Fourems Province`.

File: processing/src/main/java/org/apache/druid/query/filter/Filter.java [hunk 9]
```diff
@@ -162,4 +163,29 @@ default boolean canVectorizeMatcher()
...
+   * based on a renaming map where the key is the column to be renamed in the filter, and the value is the new
+   * column name.
+   *
+   * For example, if I have a filter (A = hello), and I have a renaming map (A -> B),
+   * this should return the filter (B = hello)
+   *
+   * @param columnRewrites Column rewrite map
+   * @return Copy of this filter that operates on new columns based on the rewrite map
+   */
+  default Filter rewriteRequiredColumns(Map<String, String> columnRewrites)
```
[the remark this answers is not in the document]
Comment (jon-wei) [reply processing/src/main/java/org/apache/druid/query/filter/Filter.java:187]: Yes, that was the intent

File: processing/src/main/java/org/apache/druid/query/filter/Filter.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (jihoonson) [reply processing/src/main/java/org/apache/druid/query/filter/Filter.java:171]: Oh thanks, I see what you are saying. But do we need that kind of tests? If I understand correctly, this interface is for temporarily marking what filters support the rewrite for now since some of filters don't implement it. This interface will be eventually removed in the future. I don't see any reason to not support the rewrite for some particular filter types.

File: processing/src/main/java/org/apache/druid/segment/join/filter/JoinFilterAnalyzer.java (hunk 8 above)
Comment (jon-wei) [reply processing/src/main/java/org/apache/druid/segment/join/filter/JoinFilterAnalyzer.java]: I've added that test and fixed the issue, the problem was I forgot to implement a block here for handling direct rewrites: https://github.com/apache/druid/pull/9818/commits/87ad5690240553200597f66185310169ad4e2b02#diff-24d942179715e417814286d06dc4f668R203

The "determine RHS candidate for leaf filter" code has been moved into a separate method, called by the change above: https://github.com/apache/druid/pull/9818/commits/87ad5690240553200597f66185310169ad4e2b02#diff-24d942179715e417814286d06dc4f668R367

As part of that I adjusted this area and related https://github.com/apache/druid/pull/9818/commits/87ad5690240553200597f66185310169ad4e2b02#diff-24d942179715e417814286d06dc4f668R229 so that the direct rewrite information is kept separately from the non-direct rewrites, sharing them was causing problems.

**Comment by jon-wei** [comment]
> I'm curious why the value type is Optional vs just treating nulls as missing keys. Are Optional.absent() and null values (missing keys) treated in different ways?

Hm, using Optional was pointless there, I adjusted to just use null here

File: processing/src/main/java/org/apache/druid/segment/filter/InFilter.java [hunk 10]
```diff
@@ -179,6 +181,30 @@ public boolean canVectorizeMatcher()
...
+
+  @Override
+  public Filter rewriteRequiredColumns(Map<String, String> columnRewrites)
+  {
+    if (columnRewrites.get(dimension) == null) {
+      throw new IAE("Received a non-applicable rewrite: %s, filter's dimension: %s", columnRewrites, dimension);
+    }
+
+    return new InFilter(
+        columnRewrites.get(dimension),
```
[the remark this answers is not in the document]
Comment (jon-wei) [reply processing/src/main/java/org/apache/druid/segment/filter/InFilter.java]: Cached the local variables in these areas

File: processing/src/main/java/org/apache/druid/segment/join/filter/JoinFilterAnalyzer.java [hunk 11]
```diff
@@ -125,7 +126,8 @@ public static JoinFilterPreAnalysis computeJoinFilterPreAnalysis(
           null,
           null,
           enableFilterPushDown,
-          enableFilterRewrite
+          enableFilterRewrite,
+          new HashMap<>()
```
[the remark this answers is not in the document]
Comment (jon-wei) [reply processing/src/main/java/org/apache/druid/segment/join/filter/JoinFilterAnalyzer.java]: Updated to suggested

File: processing/src/main/java/org/apache/druid/segment/join/filter/JoinFilterAnalyzer.java [hunk 12]
```diff
@@ -166,7 +168,8 @@ public static JoinFilterPreAnalysis computeJoinFilterPreAnalysis(
           normalizedJoinTableClauses,
           null,
           enableFilterPushDown,
-          enableFilterRewrite
+          enableFilterRewrite,
+          new HashMap<>()
```
[the remark this answers is not in the document]
Comment (jon-wei) [reply processing/src/main/java/org/apache/druid/segment/join/filter/JoinFilterAnalyzer.java]: Updated to suggested

File: processing/src/main/java/org/apache/druid/segment/join/filter/JoinFilterAnalyzer.java [hunk 13]
```diff
@@ -862,5 +1006,10 @@ public String getValueForRewrite()
     {
       return valueForRewrite;
     }
+
+    public boolean isDirectRewrite()
```
[the remark this answers is not in the document]
Comment (jon-wei) [reply processing/src/main/java/org/apache/druid/segment/join/filter/JoinFilterAnalyzer.java:1096]: Added a javadoc

File: processing/src/main/java/org/apache/druid/segment/join/filter/JoinFilterAnalyzer.java [hunk 14]
```diff
@@ -370,14 +412,15 @@ public static JoinFilterSplit splitFilter(
...
+   * @param filterClause          Individual filter clause (an OR filter or a leaf filter) from a filter that is in CNF
    * @param joinFilterPreAnalysis The pre-analysis computed by {@link #computeJoinFilterPreAnalysis)}
    *
    * @return a JoinFilterAnalysis that contains a possible filter rewrite and information on how to handle the filter.
    */
   private static JoinFilterAnalysis analyzeJoinFilterClause(
       Filter filterClause,
-      JoinFilterPreAnalysis joinFilterPreAnalysis
+      JoinFilterPreAnalysis joinFilterPreAnalysis,
+      Map<Expr, VirtualColumn> pushDownVirtualColumnsForLhsExprs
```
[the remark this answers is not in the document]
Comment (jon-wei) [reply processing/src/main/java/org/apache/druid/segment/join/filter/JoinFilterAnalyzer.java:497]: Added new parameter to javadoc

[the remark this answers is not in the document]
Comment (jihoonson) [reply processing/src/main/java/org/apache/druid/segment/join/filter/JoinFilterAnalyzer.java:497]: Thanks. This javadoc is really helpful.

File: processing/src/main/java/org/apache/druid/segment/join/filter/JoinFilterAnalyzer.java (hunk 8 above)
Comment (jihoonson) [reply processing/src/main/java/org/apache/druid/segment/join/filter/JoinFilterAnalyzer.java]: Thanks for the quick fix!
